### ISSUE TITLE
[ttnn] unary elementwise: 28 op types return float32 bit patterns for int32 input instead of raising

### ISSUE THREAD

**Issue by heroxcreed** [tenstorrent/tt-metal#56938]
### Summary

Passing an `int32` tensor to the unary elementwise ops listed below does not raise, and the output tensor still
reports `dtype == INT32` — but the stored values are the **float32 result reinterpreted as int32**. For many ops
the output is a constant for every positive input (`sqrt` → `0`, `sigmoid` → `1056964608` (0.5f),
`exp` → `1065353216` (1.0f), `log` → `-8388608` (-inf)).

Of the **43 unary op types** measured: **28** return float32 bit patterns (this issue); **5** (`floor`, `round`,
`trunc`, `hardshrink`, `mish`) return plausible-looking but wrong integers and are already described in #56853;
**7** are integer-correct (`abs`, `bitwise_not`, `identity`, `neg`, `relu`, `relu6`, `square`); the remaining
**3** (`clamp`, `clip`, `leaky_relu`) do not accept the call at all (`clamp`/`clip` raise `TT_FATAL` in the
unary composite path). So 33 of 43 are affected, and the same dtype is accepted, rejected, or silently
mishandled depending on the op type.

Measured on the Blackhole simulator (`ttnn` wheel 0.78.0).

### Reproduction

```python
import torch, numpy as np, ttnn

dev = ttnn.open_device(device_id=0)
x = torch.from_numpy(np.array([-3, -1, 0, 1, 3, 7], dtype=np.int32))
t = ttnn.from_torch(x.reshape(1, 1, 1, -1).contiguous(), dtype=ttnn.int32,
                    layout=ttnn.TILE_LAYOUT, device=dev)

out = ttnn.sign(t)
print(out.dtype)          # DataType.INT32
print(ttnn.to_torch(out)) # tensor([-1082130432, -1082130432, 0, 1065353216, 1065353216, 1065353216])
                          #   = the float32 bit patterns of -1.0f, 0.0f, +1.0f
ttnn.close_device(dev)
```

`ttnn.sqrt` on the same input returns `[2143289344, 2143289344, 0, 0, 0, 0]`, i.e. `[NaN, NaN, 0, 0, 0, 0]` read
as float — zero for every positive input. No error, no warning, `dtype` still `INT32`.

### Representative outputs (input `[-3, -1, 0, 1, 3, 7]`)

| op | stored int32 values | interpreted as float |
|---|---|---|
| `sign` | -1082130432, -1082130432, 0, 1065353216, … | -1.0, -1.0, 0, 1.0, … |
| `sqrt` | 2143289344, 2143289344, 0, 0, 0, 0 | NaN, NaN, 0, 0, 0, 0 |
| `sigmoid` | 0, 0, 1056964608, 1056964608, … | 0, 0, 0.5, 0.5, … |
| `hardsigmoid`, `hardswish` | 1065353216, 1065353216, 1056964608, … | 1.0, 1.0, 0.5, … |
| `exp`, `cos`, `erf`, `isnan`, `ceil` | …, 1065353216, … | …, 1.0, … |
| `log` | …, -8388608, … | …, -inf, … |
| `reciprocal` | -2147483648, -2147483648, 2139095040, … | -0.0, -0.0, inf, … |
| `rsqrt` | 2143289344, 2139095040, 1610110408, … | NaN, inf, 3.58e19, … |
| `softplus` | -3, -1, 1060223412, … | 0, 0, 0.6942399, … |
| `erfc` | 2143289344, 2143289344, 1065288070, … | NaN, NaN, 0.996117, … |

**Affected (28)**: `atan`, `ceil`, `celu`, `cos`, `elu`, `erf`, `erfc`, `exp`, `frac`, `gelu`, `hardsigmoid`,
`hardswish`, `hardtanh`, `isnan`, `log`, `reciprocal`, `rsqrt`, `selu`, `sigmoid`, `sign`, `silu`, `sin`,
`softplus`, `softshrink`, `softsign`, `sqrt`, `tanh`, `tanhshrink`

### Mechanism

The SFPU kernels read and write the destination register as float regardless of the tensor dtype. In
`tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sign.h` (`origin/main`), `calculate_sign` is:

```cpp
sfpi::vFloat v = sfpi::dst_reg[0];                       // the int32 payload is read as a float
sfpi::vFloat res = sfpi::copysgn(sfpi::vFloat(1.0f), v);
v_if(sfpi::abs(v) == 0.0F) { res = 0.0f; }
v_endif;
sfpi::dst_reg[0] = res;                                  // the float result is written back into the INT32 tensor
```

(The same file in the measured 0.78.0 wheel used the earlier `v_if(v < 0.0F)` / `_sfpu_is_fp16_zero_` form; the
read-float / write-float datapath is unchanged.)

In `unary/common/unary_op_utils.cpp`, `get_op_init_and_func_default` emits an integer kernel only for a subset of
op types (`RELU`, `SIGNBIT`, `LOGICAL_NOT_UNARY`, `SQUARE`, `EQZ`, `NEZ`, `LTZ`, `GTZ`, `GEZ`, `LEZ`, `RELU6`,
`NEG`, plus the `ABS_INT32` op type and the bitwise/shift/fill family). Every other op type gets a plain float
tile function — `case UnaryOpType::SIGN: return {"sign_tile_init();", …}` next to
`case UnaryOpType::ABS_INT32: … abs_tile_int32(…)` — i.e. the affected set is exactly the set of op types with
no integer branch in that switch.

### Why this reads as a defect rather than a documented limitation

* The neighbouring elementwise paths validate the same dtype: `binary_ng` raises
  `TT_FATAL @ binary_ng_device_operation.cpp:339` for the float-only op types (`atan2`, `bias_gelu`, `hypot`,
  `logaddexp`, `logaddexp2`, `pow`, `xlogy`), and ternary `addcdiv` raises `TT_FATAL: !is_input_int32`
  (`ternary.cpp:416`).
* `unary.cpp::mac_tss` raises for integer input with the comment "integer inputs cannot be served here", and
  `where_tss` typecasts an int32 condition to FLOAT32 rather than feeding it to the fp32 SFPU — both patterns
  for this exact problem already exist in the file.
* `unary/device/unary_device_operation.cpp` already carries per-op dtype rules of the same kind
  (`Unary: LGAMMA requires BFLOAT16 or FLOAT32 input`, `Unary: SOFTCAP supports BFLOAT16 and BFLOAT8_B inputs`),
  but has no int32 rule for the op types above.
* int32 is de facto supported on this path where an integer kernel exists: `unary.cpp:91-93` maps int32 `abs` to
  `UnaryOpType::ABS_INT32` (`unary_op_utils.cpp:819` emits `abs_tile_int32`), and
  `tests/ttnn/unit_tests/operations/eltwise/test_unary_int32.py` exercises int32 for
  `abs` / `bitwise_*` / `logical_*` / `square`.
* Coverage gap: `test_unary_int32.py` has no `sign` case, while `ttnn.signbit` has an explicit int32
  edge-case test.

So the same dtype is handled four different ways across the elementwise routes:

| route | behaviour on int32 |
|---|---|
| `binary_ng`, tensor-tensor (`add`/`mul`/`minimum`/`maximum`/`bitwise_*`/`gcd`/`lcm`/`gt`/`logical_and`/`nextafter`/`squared_difference`/`where`/`addcmul`/`lerp`) | accepted, integer-correct |
| `binary_ng`, float-only op types | **rejected with `TT_FATAL`** |
| ternary (`addcdiv`) | **rejected with `TT_FATAL: !is_input_int32`** |
| unary — the 28 op types above | **accepted silently, wrong `INT32` contents** |

### Suggested fix

(a) Validate the dtype on the unary elementwise path — a check next to the existing `LGAMMA`/`SOFTCAP` rules,
raising `TT_FATAL` for int32 on those op types, mirroring the `binary_ng` policy and the `mac_tss` precedent.
(b) Or typecast as `where_tss` does and return a FLOAT32 tensor, if these calls should keep working with float
semantics. (c) Adding integer kernels for all 28 is the largest change.
I can prepare the patch and the int32 tests for whichever direction you prefer.

### Environment / caveats

* Runtime numbers: Blackhole **simulator**, `ttnn` wheel **0.78.0** (the newest published wheel). **Hardware not
  verified.** The per-op values above are the raw recorded outputs of that run.
* Source statements were re-checked today against `origin/main` = `43a157fb81`. Note that the host-side files are
  not identical between the measured wheel and `main`: the sign LLK was rewritten upstream (now `copysign`-based,
  as quoted), and `unary_op_utils.cpp` is not shipped inside the wheel at all. What is verified at `main` is the
  mechanism: `get_op_init_and_func_default` still has **no int32 branch** for these 28 op types, and
  `unary_device_operation.cpp` still has no int32 rule for them.
* Re-running the measurement on `main` was **not** possible in my environment: the Blackhole simulator available
  to me (`libttsim_bh.so`) rejects an SFPU instruction emitted by current-`main` kernels
  (`UnsupportedFunctionality: tensix_sfpgt: lreg_dest=9`). So the runtime numbers above are 0.78.0, and the
  "still present on `main`" claim rests on the source check, not on a run.
* Duplicate check (2026-09-18): open+closed issues and open PRs searched for `int32`, `bit pattern`,
  `reinterpret`, `unary`, `floor`, `round`, `trunc`, `mish`, `bias_gelu`. The closest reports are #48240 (unary
  int32/**uint32 scalar-parameter** handling, a different mechanism), #55311 / #55603 (the `dtype=` argument
  being dropped, not int32 input), and #53908 (unary comparisons). No report of this behaviour was found.
* I can supply the full measurement script and the per-op raw outputs on request.

**Comment by mateusznowakTT** [comment]
@heroxcreed : Please refer to https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api.html for supported data types for ops.

**Comment by heroxcreed** [comment]
Thanks for the pointer. I am not disputing the supported dtype list - the docs are clear
that these ops target BFLOAT16 / BFLOAT8_B / FLOAT32.

What I am reporting is the silent success rather than the missing dtype support: the op
accepts an int32 tensor and returns a value that is a float32 bit pattern written into the
int32 output, with no error. The follow-up PR (#56943) adds exactly the rejection for the
op types it covers, and other elementwise routes already raise TT_FATAL for int32 input
(ttnn.bias_gelu(t, t), ttnn.pow(t, t), ttnn.addcdiv), so the validation currently lands
inconsistently.

If the intent is that unsupported dtypes should silently proceed, I will drop it; if the
intent is to reject them, I am happy to help extend the coverage to the remaining op types.
